middleware/context_timeout_test.go

```diff
@@ -0,0 +1,431 @@
+		// Timeout has to be defined or the whole flow for timeout middleware will be skipped
+		Timeout: 50 * time.Millisecond,
+	})
+
+	req := httptest.NewRequest(http.MethodGet, "/", nil)
+	rec := httptest.NewRecorder()
+
+	e := echo.New()
+	c := e.NewContext(req, rec)
+
+	err := m(func(c echo.Context) error {
+		return c.JSON(http.StatusCreated, map[string]string{"data": "ok"})
+	})(c)
+
+	assert.NoError(t, err)
+	assert.Equal(t, http.StatusCreated, rec.Code)
+	assert.Equal(t, "{\"data\":\"ok\"}\n", rec.Body.String())
+}
+
+func TestContextTimeoutTestRequestClone(t *testing.T) {
+	t.Parallel()
+	req := httptest.NewRequest(http.MethodPost, "/uri?query=value", strings.NewReader(url.Values{"form": {"value"}}.Encode()))
+	req.AddCookie(&http.Cookie{Name: "cookie", Value: "value"})
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+	rec := httptest.NewRecorder()
+
+	m := ContextTimeoutWithConfig(ContextTimeoutConfig{
+		// Timeout has to be defined or the whole flow for timeout middleware will be skipped
+		Timeout: 1 * time.Second,
+	})
+
+	e := echo.New()
+	c := e.NewContext(req, rec)
+
+	err := m(func(c echo.Context) error {
+		// Cookie test
+		cookie, err := c.Request().Cookie("cookie")
+		if assert.NoError(t, err) {
+			assert.EqualValues(t, "cookie", cookie.Name)
+			assert.EqualValues(t, "value", cookie.Value)
+		}
+
+		// Form values
+		if assert.NoError(t, c.Request().ParseForm()) {
+			assert.EqualValues(t, "value", c.Request().FormValue("form"))
+		}
+
+		// Query string
+		assert.EqualValues(t, "value", c.Request().URL.Query()["query"][0])
+		return nil
+	})(c)
+
+	assert.NoError(t, err)
+}
+
+func TestContextTimeoutRecoversPanic(t *testing.T) {
+	t.Parallel()
+	e := echo.New()
+	e.Use(Recover()) // recover middleware will handler our panic
+	e.Use(ContextTimeoutWithConfig(ContextTimeoutConfig{
+		Timeout: 50 * time.Millisecond,
+	}))
+
+	e.GET("/", func(c echo.Context) error {
+		panic("panic!!!")
+	})
+
+	req := httptest.NewRequest(http.MethodGet, "/", nil)
+	rec := httptest.NewRecorder()
+
+	assert.NotPanics(t, func() {
+		e.ServeHTTP(rec, req)
+	})
+}
+
+func TestContextTimeoutDataRace(t *testing.T) {
+	t.Parallel()
+
+	timeout := 1 * time.Millisecond
+	m := ContextTimeoutWithConfig(ContextTimeoutConfig{
+		Timeout: timeout,
+		//ErrorMessage: "Timeout! change me",
+	})
+
+	req := httptest.NewRequest(http.MethodGet, "/", nil)
+	rec := httptest.NewRecorder()
+
+	e := echo.New()
+	c := e.NewContext(req, rec)
+
+	err := m(func(c echo.Context) error {
+
+		if err := sleepWithContext(c.Request().Context(), timeout); err != nil {
+			return err
+		}
+		return c.String(http.StatusOK, "Hello, World!")
+	})(c)
+
+	assert.NoError(t, err)
+
+	if rec.Code == http.StatusServiceUnavailable {
+		assert.Equal(t, "Timeout! change me", rec.Body.String())
+	} else {
+		assert.Equal(t, "Hello, World!", rec.Body.String())
+	}
+}
+
+func TestContextTimeoutWithErrorMessage(t *testing.T) {
```

Comment:
I think copying that other `Timeout()` middleware tests is unnecessary. This variant of timeout is miles simpler and does no include things run in different coroutine.  Most of the tests that other timeout middleware has are because things are run in different coroutine.